test/Spec/HydraAuctionOnchain/Validators/StandingBid.hs

```diff
@@ -0,0 +1,684 @@
+  shouldFailWithError StandingBid'MoveToHydra'Error'MissingDelegateSignatures $
+    testMoveToHydra testContext $
+      def
+        { mh'txSignedByAllDelegates = False
+        }
+
+prop_moveToHydra_incorrectValidRange_fails :: MoveToHydraTestContext -> Property
+prop_moveToHydra_incorrectValidRange_fails testContext =
+  shouldFailWithError StandingBid'MoveToHydra'Error'IncorrectValidityInterval $
+    testMoveToHydra testContext $
+      def
+        { mh'txHasIncorrectValidRange = True
+        }
+
+-- ConcludeAuction ---------------------------------------------------
+
+prop_concludeAuction_validInput_succeeds :: ConcludeAuctionTestContext -> Property
+prop_concludeAuction_validInput_succeeds testContext =
+  shouldSucceed $
+    testConcludeAuction testContext def
+
+prop_concludeAuction_missingAuctionEscrowInput_fails :: ConcludeAuctionTestContext -> Property
+prop_concludeAuction_missingAuctionEscrowInput_fails testContext =
+  shouldFailWithError StandingBid'ConcludeAuction'Error'MissingAuctionEscrowInput $
+    testConcludeAuction testContext $
+      def
+        { ca'auctionEscrowInputMode = TxOutputMissing
+        }
+
+prop_concludeAuction_multipleAuctionEscrowInputs_fails
+  :: ConcludeAuctionTestContext -> Property
+prop_concludeAuction_multipleAuctionEscrowInputs_fails testContext =
+  shouldFailWithError StandingBid'ConcludeAuction'Error'MissingAuctionEscrowInput $
+    testConcludeAuction testContext $
+      def
+        { ca'auctionEscrowInputMode = MultipleTxOutputs
+        }
+
+prop_concludeAuction_auctionEscrowInputMissingToken_fails
+  :: ConcludeAuctionTestContext -> Property
+prop_concludeAuction_auctionEscrowInputMissingToken_fails testContext =
+  shouldFailWithError StandingBid'ConcludeAuction'Error'MissingAuctionEscrowInput $
+    testConcludeAuction testContext $
+      def
+        { ca'auctionEscrowInputContainsToken = False
+        }
+
+prop_concludeAuction_auctionEscrowWrongRedeemer_fails :: ConcludeAuctionTestContext -> Property
+prop_concludeAuction_auctionEscrowWrongRedeemer_fails testContext =
+  shouldFailWithError StandingBid'ConcludeAuction'Error'InvalidAuctionEscrowRedeemer $
+    testConcludeAuction testContext $
+      def
+        { ca'spendsAuctionEscrowInputWithWrongRedeemer = True
+        }
+
+--------------------------------------------------------------------------------
+-- TestContext
+--------------------------------------------------------------------------------
+
+-- NewBid ------------------------------------------------------------
+
+data NewBidTestContext = NewBidTestContext
+  { auctionCs :: CurrencySymbol
+  , auctionTerms :: AuctionTerms
+  , oldBidState :: StandingBidState
+  , newBidState :: StandingBidState
+  , txInfoTemplate :: TxInfo
+  , standingBidInputOref :: TxOutRef
+  , scriptAddress :: Address
+  , invalidNewBidStateDatum :: Datum
+  , invalidMintValue :: Value
+  , incorrectTxValidRange :: POSIXTimeRange
+  }
+  deriving stock (Show, Eq)
```

Comment:
I quite like these Context types for testing.

Also, I think that similar types (without the "invalid" fields) would be useful to decompose off-chain endpoints into:

1. Impure queries that populate the context
2. A pure function that turns context into a transaction
3. Tx submission, monitoring, etc.

Although, knowing you, that's probably how you already do it in Purescript land. :)